virttest/qemu_vm.py

```diff
@@ -1107,7 +1107,25 @@ def add_qxl(qxl_nr, base_addr=29):
             return qxl_str
 
         def add_vga(vga):
-            return " -vga %s" % vga
+            """Add primary vga device."""
+            pci_host_bus = {'aobject': 'pci.0'}
+            vga_dev_map = {
+                # name: (driver, parent bus)
+                "std": ("VGA", pci_host_bus),
+                "cirrus": ("cirrus-vga", pci_host_bus),
+                "vmware": ("vmware-svga", pci_host_bus),
+                "qxl": ("qxl-vga", pci_host_bus),
+                "virtio": ("virtio-vga", pci_host_bus),
```

Comment:
this will not work for mmio based architectures. How about using `"virtio": ("virtio-vga-pci", pci_host_bus)` and on line 1127 add:

```python
if '-mmio:' in params.get('machine_type', ''):
   if vga_dev.endswith('-pci'):
        vga_dev = "%s-device" % vga_dev[:-4]
        parent_bus = None
```

(or something like that... Just for record I prefer if elif elif for mapping purposes, but it's just my preference)

Reply:
OK let me revise my idea:

```python
if '-mmio:' in params.get('machine_type', ''):
    parent_bus = None
else:
    parent_bus = {'aobject': 'pci.0'}

if (params.get("vga_use_legacy_expression") == "yes":
    name = "VGA-%s" % vga
    cmdline = " -vga %s" % vga
    return StrDev(name, cmdline=cmdline, parent_bus=parent_bus)

if vga == "std":
    driver = "VGA"
elif vga == ...:
    ....
else:
    # Allow specifying custom vga drivers
    driver = vga
if '-mmio:' in params.get('machine_type', '') and driver.endswith("-pci"):
    driver = "%s-device" % driver[:-4]
return QDevice(driver, parent_bus = parent_bus)
```

Reply:
(or using `{}.get` with default to `vga`)

Reply:
Thanks for pointing out the issue, I'll have a look at that.

Reply:
I'll make it configurable for different machines, how about:
```python
default_machine_regex = re.compile(".*")
vga_dev_map = {
    # name: [(machine regex, driver, parent bus), ...]
    # the default variant must be the last one
    "virtio": [(re.compile(".*-mmio:.*"), "virtio-vga-device", None),
               (default_machine_regex, "virtio-vga-pci", pci_host_bus)],
    ...
}

vga_dev = None
parent_bus = None
for variant in vga_dev_map.get(vga, []):
    if variant[0].match(params.get('machine_type', '')):
        vga_dev = variant[1]
        parent_bus = variant[2]
        break
```

Reply:
I don't understand why do you still insist on a single mapping. There is `avocado-vt` VGA type and here we need to map it to `qemu` VGA type. Yes, for different machines the `qemu` VGA type could differ, but creating a single mapping expands the matrix and requires checking for all cases.

This is why I keep suggesting 2 mappings (or mapping + if/elif/else) where first we map it to the default PCI type and in case we are on "odd" architecture we could re-map it to the related VGA.


The difference is, that by default you only do:

```
qemu_vga_type = vga_map.get(avocado_vga_type)
```

and on "odd" architectures (mmio/isapc):

```
qemu_vga_type = vga_map.get(avocado_vga_type)
vga_type = mmio_map(qemu_vga)
```

Also the `vga_map` is fairly simple, as well as the `mmio` map (simply replace `-pci` for `-device` and you're done) and for isapci it's a simple mapping for 2 existing devices or exception)